add plugin quickstart docs page (#446)
Deltares/hydromt: docs/dev/plugin-quickstart.rst

properties and the initiliasation of your new Model subclass.

To fully initiliase your new subclass (eg *MyModelModel*), you need to initialise a couple of high level properties and in some cases, you may wish to modify
the default initialisation function of you parent class (the HydroMT core class you choose, ``Model`` or ``GridModel`` etc.).

.. TIP::

  In python, a child class (eg *MyModelModel(Model)*), inherits all methods and properties of the parent class (``Model``). If you wish to overwrite completely one
  of the parent method (eg ``setup_region``), you can just redefine the function in the child class using the name (``setup_region``). You can also decide to partly
  re-use the parent function and for example add some extra steps, change the default values or the docstring of the function. For this, you can redefine the
  function in the child class, and within the redefined function, call the parent one use *super()* python class attribute (*super().setup_region*). You will see
  an example below for the initialisation function ``__init__`` .

Below are the properties and initialisation function, to set for your new plugin *MyModelModel* class:

> .. code-block:: python

Reviewer: I wonder if we should have this here as it will require keeping it in sync with the cookiecutter. Perhaps a notebook using the cookiecutter would be more suitable for this level docs? In such notebook you can also demonstrate the inheritance of properties, available methods, etc.

Author: I agree with this sentiment but I have no idea how good the compatibility between cookiecutter and notebooks are, but I don't imagine it's great and honestly I'm not very keen on the idea. I'm not sure having it included here adds that much so I just removed it. I think people can just use the template for this if they are curious.

  from hydromt.models import Model
  from os.path import abspath, join, dirname

  class MyModelModel(Model):
    """This is the HydroMT MyModel class for MyModel models."""

    ### Any global class variables your model should have go here ###
    # shortname of your model / hydromt model name for CLI
    _NAME: str = "mymodel"
    # Default filename for mymodel configuration file / run settings
    _CONF: str = "model.yaml"
    # Path to default model building for my model / eg template configuration files to edit
    # default paremter values etc.
    # This folder in many existing plugins is usually placed at the same level of the class name
    # definition in a floder called "data".
    _DATADIR: Path = abspath(join(dirname(__file__), "data"))
    # tell hydroMT which methods should receive the res and region arguments from CLI
    _CLI_ARGS: Dict[str, str] = {"region": "setup_region", "res": None}

>     ### Dictionnary of hydromt model components mapping ###
>     # Can be used for example to rename variables in the different components from
>     # hydromt convention name to the model specific name.
>     _GEOMS: Dict[str, Any] = {}
>     _MAPS: Dict[str, Any] = {"elevtn": "dem"}
>     _FORCINGS: Dict[str, Any] = {}
>     _STATES: Dict[str, Any] = {}
>     _RESULTS: Dict[str, Any] = {}
>     # If your parent class is GridModel, MeshModel, LumpedModel or NetworkModel
>     # you can also add to the list here

Reviewer: ```suggestion
```

This part is incorrect (these are not class properties but model instance properties and initialized with lower case after initialization) and should be removed, see also https://github.com/Deltares/hydromt-plugin-template/pull/3

    ### Name of default folders to create in the model directory ###
    # When initiliasing a new model for a specific region, the folders
    # int he list will be created
    _FOLDERS: List[str] = []

    ### Name of defaults catalogs to include when initialising the model ##
    # For example to include model specific parameter data or mapping
    # These default catalogs can be placed in the _DATADIR folder.

>     _CATALOGS: List[str] = [join(_DATADIR, "parameters_data.yml")]

Reviewer: I think this is to specific, see also https://github.com/Deltares/hydromt-plugin-template/pull/3

  HydroMT defines any model through the model-agnostic Model API based on several *general* components and *computational* unit components.
  Each component represents a specific model data type and is parsed to a specific Python data object.
  The general components are **maps** (raster data), **geoms** (vector data), **forcing**, **results**, **states**, and **config**
  (the model simulation configuration). These are available to all model classes and plugins.
  The computational components are different for different types of models: i.e. **grid** for distributed or grid models,
  **response_units** for lumped or semi-distributed models, **mesh** for mesh or unstructured grid models, and **network** for
  network models (to be developed).

So typically, when you want to add data to your model (DEM, precipitation, parameters etc.), you add data to one of the
specific Model component. To add data, to a component, the ``set_<component>`` methods are you use and each of the components can be
read and written using the specific ``read_<component>`` and ``write_<component>`` methods.

The table below lists the base model components common to all model classes.

> All base model attributes and methods can be found in the :ref:`API reference <model_api>`

Reviewer: Super useful, but why not linking to the same table in the model_overview.rst
I'm a bit afraid of maintenance and clarity of the docs if the same thing is stated in (many) different places.

     - Model class
     - Explanation
     - Python object
     - API
   * - grid
     - :ref:`GridModel <grid_model_api>`
     - Static gridded data with on unified grid
     - xarray.Dataset (compatible with hydromt.raster.RasterDataset)
     - | :py:attr:`~GridModel.grid`
       | :py:func:`~GridModel.set_grid`
       | :py:func:`~GridModel.read_grid`
       | :py:func:`~GridModel.write_grid`
   * - response_units
     - :ref:`LumpedModel <lumped_model_api>`
     - Static lumped data over the response_units
     - xarray.Dataset (compatible with hydromt.vector.GeoDataset)
     - | :py:attr:`~LumpedModel.response_units`
       | :py:func:`~LumpedModel.set_response_units`
       | :py:func:`~LumpedModel.read_response_units`
       | :py:func:`~LumpedModel.write_response_units`
   * - mesh
     - :ref:`MeshModel <mesh_model_api>`
     - Static mesh (unstructured grid(s)) data
     - xugrid.UgridDataset
     - | :py:attr:`~MeshModel.mesh`
       | :py:func:`~MeshModel.set_mesh`
       | :py:func:`~MeshModel.get_mesh`
       | :py:func:`~MeshModel.read_mesh`
       | :py:func:`~MeshModel.write_mesh`

Reading & Writing model components
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Because the way that models consume data is so hyper specialised, the main part of writing a plugin is writing the

Reviewer: ```suggestion
Because the way that models consume data is often model software specific, the main part of writing a plugin is writing the
```

Input/Output (IO) methods to prepare data exactly as your model expects them, in the correct format. To do this you'll
be overwriting most ``read_<component>`` or ``write_<component>`` methods in your model class. That will ensure that the
data comes in and goes out expecting exactly as you want.

In the default HydroMT CORE read and write methods, xarray objects are written as netcdf files, ``geoms`` as GeoJSON files by

> default but any geopandas driver is allowed, and ``config`` eithey as TOML, YAML or INI file. So again, you could decide to

Reviewer: ```suggestion
default but any geopandas driver is allowed, and ``config`` either as TOML, YAML or INI file. So again, you could decide to
```

completely overwrite the parent read and write methods or change the defaults.

Eg if your geoms should be written in GeoPackage format rather than GeoJSON and in a geometry folder in your model:

.. code-block:: python

  def write_geoms(self):
    """ Write model geometries to a GeoPackage file at <root>/geometry/<name>.gpkg """

    super().write_geoms(
      fn = "geometry/{name}.gpkg",
      driver = 'GPKG'
    )

If you are writting your own read/write functions, some good pratice steps of the function:

> 1. For writting, check first if the object you are writting is not empty, then you can pass. Example:

Reviewer: ```suggestion
1. For writing, check first if the object you are writing is not empty, then you can pass. Example:
```

.. code-block:: python

  if len(self.geoms) == 0:
    self.logger.debug("No geoms data found, skip writing.")
    return

> 2. For reading or writting, check that you have the right permission. For example, in build mode ('w' or 'w+'),

Reviewer: ```suggestion
2. For reading or writing, check that you have the right permission. For example, in build mode ('w' or 'w+'),
```

the model should be empty, so you should not be able to read existing model files. In some cases, python
users may only want to read a model using HydroMT to visualise or plot data. As a protection, the model
can then be opened in read only mode ('r') and HydroMT should not be able to overwrite any data. HydroMT
`Model` class has two porperties you can use to check for this:

- `self._assert_read_mode` if a model can be read
- `self._assert_write_mode` if a model can be written
3. If needed transform, and write the component to the right format and file(s). Don't hesitate to put
some logger information for your user.

> Set model components

Reviewer: In this section the main message should be to **not** change the set methods and properties which are defined in the Model class. So also showing the code is not necessary I think. It is however good to explain that the data is read upon first call of the property

    for name in data_dict:
        if name in self.maps:  # trigger init / read
            self.logger.warning(f"Replacing result: {name}")
        self._maps[name] = data_dict[name]

Additional Model properties
^^^^^^^^^^^^^^^^^^^^^^^^^^^
You can have a look at the :ref:`Model API documentation <model_api>` to find out all other properties of the ``Model`` class and
the other subclasses. Apart from the components here are a couple of useful properties to use when developping your plugin:

- :py:attr:`~Model.root`: path to the root folder of the model instance you are working with.
- :py:attr:`~Model.crs`: the reference coordinate system (pyproj.CRS) of the model.
- :py:attr:`~Model.region`: the region (GeoDataFrame) of your model, stored in :py:attr:`~Model.geoms`.
- :py:attr:`~Model.data_catalog`: the current data catalog you can use to add data to your model.
- :py:attr:`~Model.logger`: the HydroMT logger object for your model instance.
- :py:attr:`~Model._read`: flag if the model is in read mode ('r' or 'r+' when initialising).
- :py:attr:`~Model._write`: flag if the model is in write mode ('w' or 'w+' when initialising).

> Some submodel classes can have additionnal attributes based on their additionnal components, so check out the :ref:`API reference <model_api>`.

Reviewer: ```suggestion
Some submodel classes can have additional attributes based on their additional components, so check out the :ref:`API reference <model_api>`.
```

Adding a new property or component
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> If you wish to add additionnal properties or attributes to your plugin Model subclass, you can either decide to add simple

Reviewer: ```suggestion
If you wish to add additional properties or attributes to your plugin Model subclass, you can either decide to add simple
```

attributes directly in the initiliasation function ``__init__`` or define specific properties. Example for a shortcut to the basin
vector in the geoms object:

.. code-block:: python

  @property
  def basins(self) -> gpd.GeoDataFrame:
    """Returns a basin(s) geometry as a geopandas.GeoDataFrame."""
    # Check in geoms
    if "basins" in self.geoms:
      gdf = self.geoms["basins"]

>     # Check in maps
>     elif self._MAPS["basins"] in self.maps:
>       gdf = (
>           self.maps[self._MAPS["basins"]]
>           .raster.vectorize()
>           .set_index("value")
>           .sort_index()
>       )
>       gdf.index.name = self._MAPS["basins"]
>       # Store in geoms to re-use later
>       self.set_geoms(gdf, name="basins")

Reviewer: ```suggestion
```

I would leave out this part to keep the example simple.

    else:
        self.logger.warning(
            f"Basin map {self._MAPS['basins']} not found in maps or geoms."
        )
        gdf = gpd.GeoDataFrame()
    return gdf

> In most cases, we hope that the components defined in HydroMT `Model` classes (``config``, ``geoms``, ``maps``, ``forcing``, ``states``,

Reviewer: I like this part, but think this should also go into and _advanced_ section

    if len(fns) > 0:
      for fn in fns:
        name = basename(fn).split(".")[0]
        tbl = pd.read_csv(fn)
        self.set_tables(tbl, name=name)

  def write_tables(self):
    """Write tables at <root>."""
    self._assert_write_mode
    if len(self.tables) > 0:
      self.logger.info("Writing table files.")
      for name in self.tables:
        fn_out = join(self.root, f"{name}.csv")
        self.tables[name].to_csv(fn_out, sep=",", index=False, header=True)

4. Define a set function in order to add or update data in your component.

.. code-block:: python

  def set_tables(self, df: pd.DataFrame, name:str):
    """Add table <pandas.DataFrame> to model."""
    if not (isinstance(df, pd.DataFrame) or isinstance(df, pd.Series)):
      raise ValueError("df type not recognized, should be pandas.DataFrame or pandas.Series.")
    if name in self._tables:
      if not self._write:
        raise IOError(f"Cannot overwrite table {name} in read-only mode")
      elif self._read:
        self.logger.warning(f"Overwriting table: {name}")
    self._tables[name] = df

.. _plugin_workflows:

Data processing: setup methods and workflows
--------------------------------------------

> Now let's focus on the wheel part of the schematic diagram or how to add data to your model using HydroMT, which is what it's all about!

Reviewer: ```suggestion
Now let's focus on the _gear wheels_ of the schematic diagram or how to add data to your model using HydroMT, which is what it's all about!
```

> Nothing here is really mandatory but more about conventions or tips also from experience with other plugins.

Reviewer: I think the key message here should be the three steps from you diagram that are required in each setup method, i.e., to read with the data catalog, process with workflows, set/update model properties

Reviewer: I now see this is mentioned in the next section. Perhaps change the order?

Author: I think this information is also present in other sections and can be improved when we go for the next iteration on the diagrams so I've removed it for now.

But in general (scheme below):

- To build a model we specify ``setup_<>`` methods which transform raw input data to a specific model variable(s), for instance
  the ``setup_landuse`` method in HydroMT-Wflow to map landuse classes to associated Wflow parameter grids which are part of
  the ``grid`` component.
- All public model methods may only contain arguments which require one of the following basic python types: string, numeric
  integer and float, boolean, None, list and dict types. This requirement makes it possible to expose these methods and their arguments
  via the hydromt configuration in YAML format.
- Data is exposed to each model method through the ``Model.data_catalog`` attribute which is an instance of the
  :py:class:`hydromt.DataCatalog`. Data of :ref:`supported data types <data_types>` is provided to model methods
  by arguments which end with ``_fn`` (short for filename) which refer to a source in the data catalog

>   based on the source name or a file based on the (relative) path to the file. Within a model method the data is read

Reviewer: I would suggest to skip the **_fn** convention. We don't use this anymore in SfincsModel as many users, use the python interface and pass objects which are not read by the data adapters.

  4. adds the data to the corresponding HydroMT model components.

A simplified example of what a setup function would look like for example to add a landuse grid from a raster input
data using an external workflow from hydromt core is:

.. code-block:: python

  def setup_landuse(
        self,
        landuse_fn: Union[str, Path, xr.DataArray],
    ):
        """Add landuse data variable to grid.

        Adds model layers:

        * **landuse_class** grid: data from landuse_fn

        Parameters
        ----------
        landuse_fn: str, Path, xr.DataArray
            Data catalog key, path to raster file or raster xarray data object.
            If a path to a raster file is provided it will be added
            to the data_catalog with its name based on the file basename without
            extension.
        """
        self.logger.info(f"Preparing landuse data from raster source {landuse_fn}")
        # 1. Read landuse raster data
        da_landuse = self.data_catalog.get_rasterdataset(
            landuse_fn,
            geom=self.region,
            buffer=2,

>             variables=["landuse"],

Reviewer: check naming convention. I think we often use `lulc`

Author: landuse is commonly used in the datacatalog, the only reference to `lulc` I could find was in hydromt/cli/main.py:
https://github.com/Deltares/hydromt/blob/effbf2f15e306e65e397969426d09c8d93912a8e/hydromt/cli/main.py#L273
Not sure which one you'd want

            reproject_method="mode",
        )
        # 3. Rename or transform from HydroMT to model conventions
        rmdict = {"landuse": "landuse_class"}
        # Or using a properly initialised _GRIDS
        # rmdict = {k: v for k, v in self._GRIDS.items() if k in ds_out.data_vars}
        ds_out = ds_out.rename(rmdict)

        # 4. Add to grid
        self.set_grid(ds_out)

.. NOTE::

  **Input data type of the setup method**: Typically a setup function tries to go from one type of dataset
  (landuse raster) to a HydroMT model component (landuse map in ``maps``). So it's good to make clear for your user in
  the setup function docstrings which type of input data this function can work with. You could decide to support
  several data types in one setup function but be aware that the GIS processing functions like resampling, reprojection can
  be quite different for a raster or a vector for example. So you could decide to create two setup functions that
  prepare the same data but from different type of input data (eg *setup_landuse_from_raster* and *setup_landuse_from_vector*).

Base or region setup method(s)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> When building a model from scratch, the first thing you should do is to define the region in the world your model is located in,

Reviewer: I think the `setup_grid`, `setup_mesh`, etc methods to define the geospatial representation of the model core computational unit is the first step. This typically (but not always) takes the region as an argument.

Reviewer: Perhaps this should also be moved to before the setup methods section as that's the more logical order I think.

A couple of tips if you want to define workflows:

- check out the :ref:`workflows available <workflows_api>` in HydroMT core
- avoid passing the HydroMT model class to your workflow function, but pass the required arguments directly (eg crs = self.crs,
  data = self.grid). Ideally the workflows work from common python objects like xarray or geopandas rather than with the ``Model`` class.
- if you want to do some GIS processing on ``RasterDataset`` or ``GeoDataset``, HydroMT defines a lot of useful methods. Check out the
  :ref:`Raster methods API doc <raster_api>` for RasterDataset and :ref:`GeoDataset methods API doc <geodataset_api>`. For ``GeoDataFrame``,
  the `geopandas <https://geopandas.org/en/stable/index.html>`_ library should have most of what you need (and for ``UgridDataset`` or mesh,
  the `xugrid <https://deltares.github.io/xugrid/>`_ library). For computing or deriving other variables from an input dataset, HydroMT
  contains also a couple of useful workflows for example ``flwdir`` for flow direction methods, ``basin_mask`` to derive basin shape, or
  ``stats`` to derive general, efficiency or extreme value statistics from data.

Good practices
--------------

> On top of code development for your plugin, you can also find some good practices for your plugin repository in the :ref:`Developper guide <contributing>`.

Reviewer: ```suggestion
On top of code development for your plugin, you can also find some good practices for your plugin repository in the :ref:`Developer guide <contributing>`.
```
